### PR TITLE
changes get_mean_condition_score to get_condition_stats

### PR DESCRIPTION
**Pull request by strangeties**
Benefits of this change include ...
1) fixing forsys dataframe impact score retrieval (prior to this change, only mean(condition score) is retrieved; however, impact score should be the sum(1-condition score) across all relevant raster pixels, which requires sum(condition score) and/or count(condition score) to be retrieved).
2) making it easy to understand _get_db_stats_for_plan output value (prior to this change, output=None means not-in-db while output=np.nan means in db, but the input geo has no intersection with the condition raster; after this change, output=None still means not-in-db while output={"count"=0} more clearly indicates the input geo has no intersection with the condition raster).

### CODE REVIEW

File: src/planscape/conditions/raster_utils.py [hunk 1]
```diff
@@ -15,17 +15,29 @@
...
+# Otherwise, returns ConditionStatistics.
+def _get_db_stats_for_plan(
+        plan_id, condition_id) -> ConditionStatistics | None:
     db_scores = ConditionScores.objects.filter(
         plan_id=plan_id).filter(condition_id=condition_id).all()
     if len(db_scores) > 0:
-        return db_scores[0].mean_score
-    return np.nan
+        db_score = db_scores[0]
+        return ConditionStatistics(
```
Comment (elsiel23) [inline src/planscape/conditions/raster_utils.py:36]: can we store ConditionStatistics directly in the db?

File: src/planscape/conditions/raster_utils.py [hunk 2]
```diff
@@ -57,23 +70,26 @@ def compute_condition_score_from_raster(
...
+            'get_condition_stats',
             (RASTER_TABLE, RASTER_SCHEMA, raster_name,
              RASTER_NAME_COLUMN, RASTER_COLUMN, geo.ewkb))
         fetch = cursor.fetchone()
-        if fetch is None or len(fetch) == 0:
-            return None
-        return fetch[0]
+        if fetch is None or len(fetch) != 3:
+            return ConditionStatistics({'mean': None,
+                                        'sum': 0.0,
```
Comment (elsiel23) [inline src/planscape/conditions/raster_utils.py:79]: why is sum 0.0 vs None? if that's a choice we're making, can we add a comment explaining why

**Comment by elsiel23** [comment]
What's the goal of us retrieving more stats? can you add that in the PR description?

**Comment by strangeties** [comment]
sure - I've updated the description by expanding on the two points that were made

File: src/planscape/plan/models.py [hunk 3]
```diff
@@ -128,7 +128,11 @@ class ConditionScores(models.Model):
     # Condition
     condition = models.ForeignKey(
         Condition, on_delete=models.CASCADE, null=False)
-   
-    # Mean score.
-    # This may be null if plan or project area geometry has no intersection with raster values. 
+
+    # The following are ondition statistics computed across relevant raster
```
Comment (elsiel23) [inline src/planscape/plan/models.py:132]: *condition

File: src/planscape/plan/migrations/0010_remove_project_max_cost_project_max_budget_and_more.py [hunk 4]
```diff
@@ -1,37 +1,79 @@
...
+# Generated by Django 4.1.3 on 2023-02-09 08:01
 
 from django.db import migrations, models
+import django.db.models.deletion
 
 
 class Migration(migrations.Migration):
 
     dependencies = [
-        ('plan', '0009_conditionscores'),
```
Comment (elsiel23) [inline src/planscape/plan/migrations/0010_remove_project_max_cost_project_max_budget_and_more.py:9]: are all of these migration files relevant to this PR?

**Review by elsiel23 — COMMENTED** [review]
Partial review, I think some files may not be a part of this PR's changes

**Comment by strangeties** [comment]
`I think some files may not be a part of this PR's changes`

Which files?

**Comment by elsiel23** [comment]
The migration files in plan/

**Comment by strangeties** [comment]
I see! I'll add more comments above. It was intended, though, since changing conditions/migrations/0003_get_mean_condition_score.py affected migrations that occurred after it.

Is there a benefit to maintaining the migrations' order? If so, I can re-do the migration.

File: src/planscape/conditions/raster_utils.py (hunk 1 above)
Comment (strangeties) [reply src/planscape/conditions/raster_utils.py:36]: I started working on making this return an instance of the ConditionScores model, but then realized we still need ConditionStatistics or something similar as an output to compute_condition_stats_from_raster, whose input is a geo and a raster name.

Not worth digging into for this pull request imo - more of a "polish" than "barebones" task.

File: src/planscape/conditions/raster_utils.py (hunk 2 above)
Comment (strangeties) [reply src/planscape/conditions/raster_utils.py:79]: The sum across zero raster pixels is 0.
The count of zero raster pixels is 0.
The mean is undefined - sum/count is 0/0.

File: src/planscape/plan/migrations/0010_remove_project_max_cost_project_max_budget_and_more.py (hunk 4 above)
Comment (strangeties) [reply src/planscape/plan/migrations/0010_remove_project_max_cost_project_max_budget_and_more.py:9]: kind of.
Changing conditions/migrations/0003_get_condition_stats.py meant changing all the migrations depending on it, which included everything beyond plan/migrations/0009*.

File: src/planscape/conditions/migrations/0003_get_condition_stats.py [hunk 5]
```diff
@@ -2,13 +2,17 @@
 from typing import Tuple
 
 SQL = """
-create or replace function get_mean_condition_score(
+create or replace function get_condition_stats(
```
[the remark this answers is not in the document]
Comment (strangeties) [reply src/planscape/conditions/migrations/0003_get_condition_stats.py:22]: Done.

File: src/planscape/conditions/raster_utils.py [hunk 6]
```diff
@@ -15,17 +15,29 @@
...
+# Returns None if no statistics are available.
+# Otherwise, returns ConditionStatistics.
+def _get_db_stats_for_plan(
+        plan_id, condition_id) -> ConditionStatistics | None:
     db_scores = ConditionScores.objects.filter(
         plan_id=plan_id).filter(condition_id=condition_id).all()
     if len(db_scores) > 0:
-        return db_scores[0].mean_score
-    return np.nan
+        db_score = db_scores[0]
```
[the remark this answers is not in the document]
Comment (strangeties) [reply src/planscape/conditions/raster_utils.py]: Done.

File: src/planscape/conditions/raster_utils.py [hunk 7]
```diff
@@ -15,17 +15,29 @@
...
+class ConditionStatistics(TypedDict):
+    # Mean across stand values (sum / count)
+    mean: float
+    # Sum of stand values.
+    sum: float
+    # The number of stands counted.
+    count: int
+
+
+# Returns None if no statistics are available.
```
[the remark this answers is not in the document]
Comment (strangeties) [reply src/planscape/conditions/raster_utils.py]: Done.

File: src/planscape/plan/models.py (hunk 3 above)
Comment (strangeties) [reply src/planscape/plan/models.py]: Done.

File: src/planscape/conditions/raster_utils.py (hunk 2 above)
Comment (elsiel23) [reply src/planscape/conditions/raster_utils.py:78]: can we add this/what you edited in the PR description to the function comment?

**Comment by elsiel23** [comment]
Thanks wendy! Maybe we should start a Wiki page on the forsys code? I think there are some things we could explain like how we calculate and define priority treatment impact and other things that would be nice to have there

**Review by elsiel23 — APPROVED** [review]

**Comment by strangeties** [comment]
Thanks.
yes - I can start looking into adding a few forsys details to the wiki.

**Pull request merged**